### ISSUE TITLE
filter_duplicate_text - Words that appear in duplicated texts have lower scores than they should

### ISSUE THREAD

**Issue by eranhirs** [elastic/elasticsearch#42780]
**Elasticsearch version** (`bin/elasticsearch --version`): [Docker Image](https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docker.html)

**Plugins installed**: []

**JVM version** (`java -version`): [Docker Image](https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docker.html)

**OS version** (`uname -a` if on a Unix-like system): Ubuntu 18

**Description of the problem including expected versus actual behavior**:

When setting `"filter_duplicate_text": true` in `significant_text` aggregation, it changes the count only for the `doc_count` and not for the `bg_count`.

This is problematic as a word that appears a lot in duplicated texts has a significantly lower score than it should.

The expected behavior would be that it ignores duplicated texts completely like they don't exist.

**Steps to reproduce**:

In the following example, we expect the words `Cat1` and `Cat2` to show up together.
However, we will see that `Cat2` is penalized because it appears in duplicated texts.
Then we remove the duplicated texts and see that `Cat2` appears like expected.

```
PUT some_index

POST some_index/_search

POST some_index/_doc/1
{
  "text": "Cat1 Cat2 - I am a duplicated text"
}

POST some_index/_doc/2
{
  "text": "Cat1 Cat2 - I am a duplicated text"
}

POST some_index/_doc/3
{
  "text": "Cat1 Cat2 - I am a duplicated text"
}

POST some_index/_doc/4
{
  "text": " not - Cat1 - duplicated - Cat2 - at all"
}

POST some_index/_doc/5
{
  "text": "no - Cat1 - duplication - Cat2"
}

POST some_index/_doc/6
{
  "text": "zero - Cat1 - dups - Cat2"
}

POST some_index/_doc/7
{
  "text": "Very different text"
}
```

When using the following search, you can see that `Cat2` is not a word that is associated with `Cat1`
```
POST some_index/_search
{
  "size": 0,
  "query": {
    "match": {
      "text": "Cat1"
    }
  }, 
  "aggs": {
    "NAME": {
      "significant_text": {
        "field": "text",
        "filter_duplicate_text": true,
        "min_doc_count": 2
      }
    }
  }
}
```

Now, delete the index and run it again without indexing the duplicated texts (remove documents 1 and 2).
Run the search again, we see Cat2 appears.

I couldn't reproduce it but in my own example I did get Cat2 but with a low score and I saw that bg_count is the same and doc_count is lower.

**Comment by markharwood** [comment]
Hi Eran,
>When setting "filter_duplicate_text": true in significant_text aggregation, it changes the count only for the doc_count and not for the bg_count.

Yes, ideally filtering duplicate text sequences the reported stats would take effect on foreground and background texts but in reality this is not practical. When sampling the top 200 or so docs for a foreground it is practical to spot duplicate text sequences as similar docs rank similarly and are likely to be bunched together in the relatively small sample. In the background (_all_ docs in the index) similar docs may be nowhere near each other - possibly on different machines, so it is not practical to find duplicate sequences of text in a space this large.

So in summary, this is working-as-designed, subject to some known limitations for which there are no good workarounds.
Closing, but if you have any brilliant ideas, we are more than happy to take another look :)

**Comment by eranhirs** [comment]
@markharwood thanks for answering so quickly .
I think solving this issue requires a smaller solution than the ideal solution that you've described.

This issue is specifically about the documents that were already filtered out from the foreground but not from the background.
We've already identified them as duplicated, why can't we just remove from the background the same count of documents that were removed from the foreground?
It would already improve the accuracy a lot.

What are your thoughts?

**Comment by markharwood** [comment]
>This issue is specifically about the documents that were already filtered out

That's perhaps where there's a disconnect. We filter _sections of documents_, not whole documents.
See https://youtu.be/zH7bizwjj20?t=236

**Comment by eranhirs** [comment]
Thank you that was very helpful.

It still sounds possible to implement a simple solution - my thoughts are an algorithm that counts the `doc_count` before and the `doc_count` after the dedup, calculates a diff and removes this diff from `bg_count`.

Quickly adding the motivation for this issue:
There are websites with `"About bird flu"` sections which is duplicated through different `Bird flu` articles.
This is a problem because it has good words like `h5n1`, but they are getting penalized and then don't count as `significant_terms`.

**Comment by markharwood** [comment]
>It still sounds possible to implement a simple solution

I'm not sure. Let's imagine a doc with `content` and `sidebar` sections - the content is the unique stuff, the `sidebar` is your `"About bird flu"` example which has useful word `h5n1` and crappy word `foobar.com`
Significant text (with filtering on) only considers words that survive the de-dup - otherwise we'd tune into the `foobar.com` in the sidebar.  So we'd consider `h5n1` and `avian` from the content.
Let's say the sidebar did not include the word `avian` - that would get a "true" background count (assuming there's no other copy-and-paste sidebars bumping up the numbers).
In contrast, the background-count for `h5n1` though, you suspect might need adjustment. Well the only adjustment we can make is the number of documents we saw in the foreground where `h5n1` was in the sidebar but not in the content.  
Implementing that will be challenging - right now filtering happens in the text-analysis phase before we get to the aggregation of terms and counts. This would be increased complexity, response times and RAM cost for what I suspect will be a small increase in usefulness in some cases. There are several other reasons why counts are incorrect and they are all performance trade-offs of this nature.
